exporters/prometheus/exporter_test.go

```diff
@@ -672,3 +676,164 @@ func TestDuplicateMetrics(t *testing.T) {
 		})
 	}
 }
+
+func TestConcurrentCollect(t *testing.T) {
+	registry := prometheus.NewRegistry()
+	cfg := newConfig(WithRegisterer(registry))
+
+	reader := metric.NewManualReader(cfg.manualReaderOptions()...)
+
+	collector := &collector{
+		reader:            reader,
+		disableTargetInfo: false,
+		withoutUnits:      true,
+		disableScopeInfo:  cfg.disableScopeInfo,
+		scopeInfos:        make(map[instrumentation.Scope]prometheus.Metric),
+		metricFamilies:    make(map[string]*dto.MetricFamily),
+	}
+
+	err := cfg.registerer.Register(collector)
+	require.NoError(t, err)
+
+	ctx := context.Background()
+
+	// initialize resource
+	res, err := resource.New(ctx,
+		resource.WithAttributes(semconv.ServiceName("prometheus_test")),
+		resource.WithAttributes(semconv.TelemetrySDKVersion("latest")),
+	)
+	require.NoError(t, err)
+	res, err = resource.Merge(resource.Default(), res)
+	require.NoError(t, err)
+
+	exporter := &Exporter{Reader: reader}
+
+	// initialize provider
+	provider := metric.NewMeterProvider(
+		metric.WithReader(exporter),
+		metric.WithResource(res),
+	)
+
+	// initialize two meter a, b
+	meterA := provider.Meter("ma", otelmetric.WithInstrumentationVersion("v0.1.0"))
+	meterB := provider.Meter("mb", otelmetric.WithInstrumentationVersion("v0.1.0"))
+
+	fooA, err := meterA.Int64Counter("foo",
+		otelmetric.WithUnit("By"),
+		otelmetric.WithDescription("meter counter foo"))
+	assert.NoError(t, err)
+
+	opt := otelmetric.WithAttributes(
+		attribute.Key("A").String("B"),
+	)
+
+	fooA.Add(ctx, 100, opt)
+
+	fooB, err := meterB.Int64Counter("foo",
+		otelmetric.WithUnit("By"),
+		otelmetric.WithDescription("meter counter foo"))
+	assert.NoError(t, err)
+	fooB.Add(ctx, 100, opt)
+
+	concurrencyLevel := 100
+	ch := make(chan prometheus.Metric, concurrencyLevel)
+
+	for i := 0; i < concurrencyLevel; i++ {
+		go func() {
```

Comment:
We should use a `var wg sync.WaitGroup` to make sure that all goroutines have finished.

Reply:
https://github.com/open-telemetry/opentelemetry-go/issues/4270